Broadcasting/remote endpoint

blazpie opened the pull request:
PR adds Remote Endpoint designed to link two RTC Engines on different nodes, and enables to share room between jellyfish instances.

At the moment it's a WIP and contains:
1. Creating remote endpoint (also linking with existing remote endpoint)
2. Concept of communication between endpoints.

![RE](https://github.com/jellyfish-dev/membrane_rtc_engine/assets/27741856/e1d9758d-56d8-42b1-abc5-f12422f61ec7)

mickel8 reviewed (changes requested):
That's not a big deal but could we take more care of commit names? :smiley: 

And the other thing, looks like you committed build files and that's a big deal as my browser cannot handle so many changes :smiley:

blazpie commented:
> That's not a big deal but could we take more care of commit names? 😃
> 
> And the other thing, looks like you committed build files and that's a big deal as my browser cannot handle so many changes 😃



> That's not a big deal but could we take more care of commit names? 😃
> 
> And the other thing, looks like you committed build files and that's a big deal as my browser cannot handle so many changes 😃

I've squashed commits and removed build files (I've got no idea how they get there as they're not related with my work)

Rados13 commented on remote/CHANGELOG.md:
```diff
@@ -0,0 +1,15 @@
+# Changelog
+
+## 0.3.0-dev
+* Fix segment duration and add framerate to hls playlist [#316](https://github.com/jellyfish-dev/membrane_rtc_engine/pull/316)
+
+## 0.2.0
+* Rename FileEndpoint to FileSourceEndpoint [#300](https://github.com/jellyfish-dev/membrane_rtc_engine/pull/300/)
+* Move StaticTrackSender to membrane_rtc_engine [#300](https://github.com/jellyfish-dev/membrane_rtc_engine/pull/300/)
+* Update HLS to use new H264 parser. Bump deps [#309](https://github.com/jellyfish-dev/membrane_rtc_engine/pull/309)
+* Changes related to RTSP->WebRTC support [#311](https://github.com/jellyfish-dev/membrane_rtc_engine/pull/311)
+* Support for any number of video tracks in HLS Compositor layout [#312](https://github.com/jellyfish-dev/membrane_rtc_engine/pull/312)
+* Set `sc_threshold=0` in H264 encoder [#313](https://github.com/jellyfish-dev/membrane_rtc_engine/pull/313)
```
Probably it should be deleted.

Rados13 commented on remote/README.md:
````diff
@@ -0,0 +1,28 @@
+<!-- # Membrane RTC Engine HLS Endpoint
+
+[![Hex.pm](https://img.shields.io/hexpm/v/membrane_rtc_engine_hls.svg)](https://hex.pm/packages/membrane_rtc_engine_hls)
+[![API Docs](https://img.shields.io/badge/api-docs-yellow.svg?style=flat)](https://hexdocs.pm/membrane_rtc_engine_hls)
+[![codecov](https://codecov.io/gh/jellyfish-dev/membrane_rtc_engine/branch/master/graph/badge.svg?token=9F1XHHUY2B)](https://codecov.io/gh/jellyfish-dev/membrane_rtc_engine)
+[![CircleCI](https://circleci.com/gh/jellyfish-dev/membrane_rtc_engine.svg?style=svg)](https://circleci.com/gh/jellyfish-dev/membrane_rtc_engine)
+
+HLS Endpoint for the [Membrane RTC Engine](https://github.com/jellyfish-dev/membrane_rtc_engine)
+
+## Installation
+
+The package can be installed by adding `membrane_rtc_engine_hls` to your list of dependencies in `mix.exs`:
+
+```elixir
+def deps do
+  [
+    {:membrane_rtc_engine_hls, "~> 0.2.0"}
+  ]
+end
+```
+
+## Copyright and License
+
+Copyright 2023, [Software Mansion](https://swmansion.com/?utm_source=git&utm_medium=readme&utm_campaign=membrane_rtc_engine)
+
+[![Software Mansion](https://logo.swmansion.com/logo?color=white&variant=desktop&width=200&tag=membrane-github)](https://swmansion.com/?utm_source=git&utm_medium=readme&utm_campaign=membrane_rtc_engine)
+
+Licensed under the [Apache License, Version 2.0](LICENSE) -->
````
Shouldn't it be updated?

Rados13 commented on remote/lib/remote_endpoint.ex:
```diff
@@ -0,0 +1,170 @@
+                spec: pid(),
+                description: "Pid of parent Engine"
+              ],
+              owner: [
+                spec: pid(),
+                description: """
+                Pid of parent all notifications will be send to.
+                These notifications are:
+                * `{:playlist_playable, content_type}`
+                * `{:cleanup, clean_function}`
+                """
+              ],
```
Isn't it a leftover from copy+paste?

Rados13 commented on remote/lib/remote_endpoint.ex:
```diff
@@ -0,0 +1,170 @@
+  @impl true
+  def handle_setup(ctx, state) do
+    {:endpoint, endpoint_id} = ctx.name
+    myself = :"#{endpoint_id}"
```
Are you sure creating atoms dynamically from UUID is a good idea? Isn't it a memory leak?

Rados13 commented on remote/lib/remote_endpoint.ex:
```diff
@@ -0,0 +1,170 @@
+    {[], state}
+  end
+
+  @impl true
+  def handle_parent_notification({:new_endpoint, _endpoint}, _ctx, state) do
+    {[], state}
+  end
+
+  @impl true
+  def handle_parent_notification({:endpoint_removed, _endpoint_id}, _ctx, state) do
+    {[], state}
+  end
+
+  @impl true
+  def handle_parent_notification({:new_tracks, _tracks}, _ctx, state) do
+    {[], state}
+  end
+
+  @impl true
+  def handle_parent_notification({:remove_tracks, _list}, _ctx, state) do
+    {[], state}
+  end
```
Why didn't we forward these messages to other rooms?

Rados13 commented on remote/lib/remote_endpoint.ex:
```diff
@@ -0,0 +1,170 @@
+    myself = :"#{endpoint_id}"
+    Process.register(self(), myself)
+
+    {actions, state_update} =
+      case state.link_proposal do
+        %Remote.LinkProposal{token: token, link_to: partner_endpoint} ->
+          token = token || generate_token()
+
+          unless is_nil(partner_endpoint) do
+            send(
+              partner_endpoint,
+              {:handshake, %Remote.LinkProposal{token: token, link_to: {myself, Node.self()}},
+               token}
+            )
+          end
+
+          {
+            [setup: :incomplete],
+            %{token: token, partner_endpoint: partner_endpoint, status: :linking}
+          }
+
+        nil ->
+          token = generate_token()
+
+          setup = %Remote.LinkProposal{
+            token: token,
+            link_to: {myself, Node.self()}
+          }
+
+          {
+            [setup: :incomplete, notify_parent: {:forward_to_parent, {:link_proposal, setup}}],
+            %{token: token, partner_endpoint: nil, status: :waiting_for_partner}
+          }
+      end
```
I don't understand why we need these lines, especially this token.

Rados13 commented on remote/lib/remote_endpoint.ex:
Doesn't this implementation require us to spawn one RemoteEndpoint to connect every two jellyfish, which means that if we have room distributed on 5 nodes each node will have 4 `remoteEndpoints`?

Rados13 reviewed (changes requested):
I am not sure if we understand the design doc I create the same way. Also, this project lacks tests and CI is not updated so it would run lint, test etc for this mix project.

mickel8 commented:
I am closing as we changed our priorities
